### PR TITLE
improve write logic to reduce cpu usage

### PR DESCRIPTION
**Pull request by hicqu**
* Avoid copy when compressing log batches;
* Calculate encoded_size during encoding a log batch;

### CODE REVIEW

File: src/engine.rs [hunk 1]
```diff
@@ -165,11 +165,17 @@ where
             loop {
                 match LogBatch::from_bytes(&mut buf, file_num, offset) {
                     Ok(Some(mut log_batch)) => {
-                        let entries_size = log_batch.entries_size();
+                        let mut encoded_size = 0;
+                        for item in &log_batch.items {
```
Comment (zhangjinpeng1987) [inline src/engine.rs:169]: Does this can reduce CPU usage?

Comment (hicqu) [reply src/engine.rs:169]: It does same thing as the old `entries_size` function.

File: src/memtable.rs [hunk 2]
```diff
@@ -444,7 +441,7 @@ impl<E: Message + Clone, W: EntryExt<E>> MemTable<E, W> {
             .rev()
             .find(|e| e.file_num <= latest_rewrite);
         if let (Some(begin), Some(end)) = (begin, end) {
-            if begin.index < end.index {
+            if begin.index <= end.index {
```
Comment (Fullstop000) [inline src/memtable.rs:444]: What's the corner case of using `<=` instead of `<`

**Review by Fullstop000 — APPROVED** [review]

File: src/memtable.rs (hunk 2 above)
Comment (hicqu) [reply src/memtable.rs:444]: If `begin.index == end.index`, there is one and only one log entry in the append queue. If we don't rewrite it, the append queue can't be purged on time. I meet the bug yesterday.

**Pull request merged**